IOMC/ParticleGuns/src/FlatRandomPtAndDxyGunProducer.cc

```diff
@@ -0,0 +1,168 @@
+
+#include "CLHEP/Random/RandFlat.h"
+
+using namespace edm;
+using namespace std;
+
+FlatRandomPtAndDxyGunProducer::FlatRandomPtAndDxyGunProducer(const ParameterSet& pset) :
+BaseFlatGunProducer(pset)
+{
+  ParameterSet defpset ;
+  ParameterSet pgun_params =
+    pset.getParameter<ParameterSet>("PGunParameters") ;
+
+  fMinPt = pgun_params.getParameter<double>("MinPt");
+  fMaxPt = pgun_params.getParameter<double>("MaxPt");
+  dxyMin_ = pgun_params.getParameter<double>("dxyMin");
+  dxyMax_ = pgun_params.getParameter<double>("dxyMax");
+  lxyMax_ = pgun_params.getParameter<double>("LxyMax");
+  lzMax_ = pgun_params.getParameter<double>("LzMax");
+  ConeRadius_ = pgun_params.getParameter<double>("ConeRadius");
+  ConeH_ = pgun_params.getParameter<double>("ConeH");
+  DistanceToAPEX_ = pgun_params.getParameter<double>("DistanceToAPEX");
+
+  produces<HepMCProduct>("unsmeared");
+  produces<GenEventInfoProduct>();
+}
+
+FlatRandomPtAndDxyGunProducer::~FlatRandomPtAndDxyGunProducer()
+{
+    // no need to cleanup GenEvent memory - done in HepMCProduct
+}
+
+void FlatRandomPtAndDxyGunProducer::produce(Event &e, const EventSetup& es)
+{
+   edm::Service<edm::RandomNumberGenerator> rng;
+   CLHEP::HepRandomEngine* engine = &rng->getEngine(e.streamID());
+
+  if ( fVerbosity > 0 ){
+    cout << " FlatRandomPtAndDxyGunProducer : Begin New Event Generation" << endl ;
+  }
+  // event loop (well, another step in it...)
+
+  // no need to clean up GenEvent memory - done in HepMCProduct
+  //
+
+  // here re-create fEvt (memory)
+  //
+  fEvt = new HepMC::GenEvent() ;
+
+  // now actualy, cook up the event from PDGTable and gun parameters
+  int barcode = 1 ;
+  for (unsigned int ip=0; ip<fPartIDs.size(); ++ip){
+
+    double phi_vtx = 0;
+    double dxy = 0;
+    double pt = 0;
+    double eta = 0;
+    double px = 0;
+    double py = 0;
+    double pz = 0;
+    double vx = 0;
+    double vy = 0;
+    double vz = 0;
+    double lxy = 0;
+
+    bool passLoop = false;
+    while (not passLoop) {
+
+      bool passLxy = false;
+      bool passLz = false;
+      phi_vtx = CLHEP::RandFlat::shoot(engine, fMinPhi, fMaxPhi);
+      dxy = CLHEP::RandFlat::shoot(engine, dxyMin_, dxyMax_);
+      float dxysign = CLHEP::RandFlat::shoot(engine, -1, 1);
+      if (dxysign < 0)
+	  dxy = -dxy;
+      pt = CLHEP::RandFlat::shoot(engine, fMinPt, fMaxPt);
+      px = pt*cos(phi_vtx);
+      py = pt*sin(phi_vtx);
+      for (int i=0; i<10000; i++){
+        vx = CLHEP::RandFlat::shoot(engine, -lxyMax_, lxyMax_);
+        vy = (pt*dxy + vx * py)/px;
+        lxy = sqrt(vx*vx + vy*vy);
+        if (lxy<abs(lxyMax_) and (vx*px + vy*py)>0){
+          passLxy = true;
+          break;
+        }
+      }
+      eta = CLHEP::RandFlat::shoot(engine, fMinEta, fMaxEta);
+      pz = pt * sinh(eta);
+      //vz = fabs(fRandomGaussGenerator->fire(0.0, LzWidth_/2.0));
+      float ConeTheta = ConeRadius_/ConeH_;
+      for (int j=0; j<100; j++){
```

Comment:
I'm not sure if we explicitly checked how many times a new vz needs to be thrown in order for it to be a good one. @tahuang1991: do you have an idea?